Fix route / isOffRoute race condition

Guardiola31337 opened the pull request:
## Description

Fixes `route` / `isOffRoute` race condition and remove unnecessary `Mutex` from `MapboxTripSession`

Fixes #3383 
Follow up from https://github.com/mapbox/mapbox-navigation-android/pull/3322#discussion_r465187851

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Deliver up to date status to clients even after they have cleared the route

### Implementation

**Edited**

~- Guard `reroute` / `isOffRoute` around `route` state - that's the best "mutex" we can use~ 😅 🚀 
- Implement a list of `Job`s for `updateNavigatorStatusData` work which are canceled (`cancelOngoingUpdateNavigatorStatusDataJobs`) as part of `route` setter and add https://github.com/mapbox/mapbox-navigation-android/blob/11b71f3172a92aec46d46e8903b2238a0351e03b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L414-L416 checks to short circuit pushing out of date updates if the job is inactive (canceled)
- Remove unnecessary `Mutex` from `MapboxTripSession` as it's not necessary anymore as above-mentioned 🎉 

Noting that getting the lock from different dispatchers (as we were) affects the overall order of the calls as it's a `suspend` function.

There is an open question as if we want to send a "last" `OffRouteObserver` notification callback or not as part of https://github.com/mapbox/mapbox-navigation-android/blob/06575cdd7cb3df8b02110329527e9cf207838f2d/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L77 Currently is sent but if we opt to not doing so we should add non-null `route` check here https://github.com/mapbox/mapbox-navigation-android/blob/06575cdd7cb3df8b02110329527e9cf207838f2d/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L111 What do you think @LukasPaczos @kmadsen?

## Testing

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

kmadsen commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:
```diff
@@ -410,7 +404,7 @@ internal class MapboxTripSession(
             launch {
```
what is this needed for again?

Guardiola31337 replied:
We had a ticket to investigate that https://github.com/mapbox/mapbox-navigation-android/issues/3265 (brought by https://github.com/mapbox/mapbox-navigation-android/pull/3255/files#r447617327 discussion).

From my testing it's needed to make sure that those don't block to each other - as they're executed in different dispatchers (`IO` and `Main`).

In any case, if you have time and can dig in deeper that'd be 💯 The more :eyes: the better 🍍

kmadsen commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:
```diff
@@ -410,7 +404,7 @@ internal class MapboxTripSession(
                 updateEnhancedLocation(status.enhancedLocation, status.keyPoints)
                 updateRouteProgress(status.routeProgress)
-                mutex.withLock {
+                route?.let {
```
does this mean `isOffRoute` is undefined when the route is null?

should isOffRoute be nullable?

Guardiola31337 replied:
> does this mean `isOffRoute` is undefined when the route is null?

Nope, `false`

> should isOffRoute be nullable?

Can be, but wouldn't that complicate the logic a bit? As we'd need to handle a new "third" state. We could use that for 👇 

> There is an open question as if we want to send a "last" `OffRouteObserver` notification callback or not as part of https://github.com/mapbox/mapbox-navigation-android/blob/06575cdd7cb3df8b02110329527e9cf207838f2d/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L77 Currently is sent but if we opt to not doing so we should add non-null `route` check here https://github.com/mapbox/mapbox-navigation-android/blob/06575cdd7cb3df8b02110329527e9cf207838f2d/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L111 What do you think @LukasPaczos @kmadsen?

although not sure if worth it as either way we'd need a check (`isOffRoute != null` vs `route != null`) so wouldn't add that much value ¯\\\_(ツ)\_/¯

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -608,7 +608,7 @@ class MapboxNavigation(
     private fun createInternalOffRouteObserver() = object : OffRouteObserver {
         override fun onOffRouteStateChanged(offRoute: Boolean) {
-            if (offRoute) {
+            if (offRoute && tripSession.route != null) {
```
`offRoute && tripSession.route == null` is a bug, we shouldn't need to keep this if-check if the rest of the logic works correctly. That's why we need to synchronize the route setter and navigator status getter (and all of the observer updates). Unless I'm missing something?

LukasPaczos replied:
What I'm trying to say is that the issue described in https://github.com/mapbox/mapbox-navigation-android/issues/3383 can still happen if I'm not mistaken:
```
1. getStatus called on a background thread and reports off-route
2. developer calls setRoutes(empty) on the main thread
3. status from 1. is switched to the main thread and delivered to the developer
```

LukasPaczos replied:
Even if `offroute` is guarded, the rest of the progress updates are still delivered, even though the developer has cleared the route already.

Guardiola31337 commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSessionTest.kt:
```diff
@@ -501,6 +614,25 @@ class MapboxTripSessionTest {
+
+        tripSession.route = null
+
+        verify(exactly = 2) { tripSession["cancelOngoingUpdateNavigatorStatusDataJobs"]() }
```
Not a big fan of testing `private` functions but that was the "better" way I found to add a test around this weird logic https://github.com/mapbox/mapbox-navigation-android/blob/11b71f3172a92aec46d46e8903b2238a0351e03b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L77-L81 added to prevent sending outdated callbacks to clients if `updateDataFromNavigatorStatus` is called multiple times consecutively.

Tried different things but couldn't `cancel` `updateNavigatorStatusDataJob` jobs externally from `TripSession` `public` APIs - without making `updateNavigatorStatusDataJobs` `internal`. Can you think of a different approach @kmadsen @LukasPaczos?

LukasPaczos replied:
We'd have to mock the jobs that are created and verify that `Job#cancel` was called.

Guardiola31337 replied:
> We'd have to mock the jobs that are created and verify that `Job#cancel` was called.

As above-mentioned, we don't have access to the `updateNavigatorStatusDataJobs` list unless we make it `internal` just for testing ¯\\\_(ツ)\_/¯ It's another option but I liked this other better because it doesn't affect "production" code.

Guardiola31337 commented:
Updated the PR as per internal discussion. This is ready for another round of 👀 @kmadsen @LukasPaczos 

Follow up work (not blocking this PR):

- [ ] Split out `ioJobController` having a dispatcher dedicated for NN
- [ ] Call `updateDataFromNavigatorStatus` right after `setRoute`

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:
```diff
@@ -70,20 +69,24 @@ internal class MapboxTripSession(
+
+    private fun cancelOngoingUpdateNavigatorStatusDataJobs() {
+        updateNavigatorStatusDataJobs.forEach {
+            it.cancel()
```
Can each element remove itself when it's canceled? Otherwise, this list might grow to thousands of elements.

We could use a synchronized collection with an iterator instead of the `CopyOnWriteArrayList` to resolve the removal and the concurrent modification exception.

Guardiola31337 replied:
> Can each element remove itself when it's canceled? Otherwise, this list might grow to thousands of elements.

Good point 👍 

> We could use a synchronized collection with an iterator instead of the `CopyOnWriteArrayList` to resolve the removal and the concurrent modification exception.

Tried that in the first place and run into `ConcurrentModificationException` as well 😅 https://stackoverflow.com/questions/25113987/why-is-there-a-concurrentmodificationexception-even-when-list-is-synchronized

LukasPaczos replied:
Oh yea, rather synchronize the access to the collection, not use a "synchronized collection", my bad :D

LukasPaczos replied:
> Can each element remove itself when it's canceled?

It should also remove itself when the job finishes I guess.

Guardiola31337 commented:
@Guardiola31337 Addressed feedback. [`invokeOnCompletion`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/invoke-on-completion.html) FTW 🚀 

Ready for another round @LukasPaczos
